### DISCUSSION TITLE
Make it possible to expand to the output of function

### DISCUSSION THREAD

**Discussion by paldepind** [olets/zsh-abbr#95] [Ideas]
In Fish it is possible to create an abbreviation that does not expand to a fixed string, but instead to the dynamic output of a function. Here is an example:
```
function last_history_item
    echo $history[1]
end
abbr -a !! --function last_history_item
```
With this abbreviation the text `!!` will expand into the most recently typed history entry.

As far as I can see this feature is not available in zsh-abbr. It's very useful, so I propose adding it to zsh-abbr.

I'm currently trying to port the command line utility [projectdo](https://github.com/paldepind/projectdo) to zsh, and it seems that I would need this feature in `zsh-abbr` to do so. If anyone knows of another way to achieve this in zsh I'd be interested to hear about it.

**Comment by olets** [comment]
> As far as I can see this feature is not available in zsh-abbr. It's very useful, so I propose adding it to zsh-abbr.

I agree the fish abbr expansion feature is cool. I don't have plans to add it to zsh-abbr.

>  If anyone knows of another way to achieve this in zsh I'd be interested to hear about it.

Me too! Maybe you can parse the `history` command? If you come up with a solution please share it.

**Comment by qadzek** [comment]
First of all, @olets, thanks a lot for this plugin. I use it all the time. It's so much beter than aliases.

@paldepind, regarding expanding `!!`, `setopt BANG_HIST` and `setopt HIST_VERIFY` is not what you are looking for?

I am looking for something similar, to expand to the output of a command. I want to expand `cdate` to the current date, so I could type `cat cdate<space>.log` to view today's log file.

I managed to do it in a hacky way by having a cron job regulary run a script that updates the abbreviation. I don't know if there is a better way?

By the way, it uses the new cursor placement feature, which is very handy for 'eating' the trailing space.

<details>
<summary>script</summary>

```bash
#!/usr/bin/env bash

cur_date="$(date +'%F')"
abbr_file="${HOME}/.config/zsh-abbr/user-abbreviations"

sed -i '/cdate/d' "$abbr_file"
echo 'abbr -g "cdate"="'"$cur_date"'%"' >> "$abbr_file"

# Add cron job if not present yet
if ! grep -qs 'insert-current-date' < <(crontab -l); then
  # shellcheck disable=2016
  cron_line='0 * * * * $HOME/dotfiles/zsh/insert-current-date-abbreviation.sh'
  (crontab -l 2>/dev/null; echo "$cron_line") | crontab -
  printf "Added cron job to regularly replace 'cdate' abbreviation\n"
fi

exit 0
```
</details>

**Reply by olets** [reply]
> thanks a lot for this plugin. I use it all the time.
> the new cursor placement feature, which is very handy

Great to hear, my pleasure!

> to expand to the output of a command

Is there a reason you can't use the `date` command in the expansion itself?

Sticking with your current model, 

```shell
# zshrc
ABBR_SET_EXPANSION_CURSOR=1
ABBR_LINE_CURSOR_MARKER=%ABBR_CURSOR% # or anything other than its default `%`

# terminal
abbr cdate="\$(date +'%F')%ABBR_CURSOR%"
cat cdate[SPACE].log
```

Fwiw in that model I'd personally pick something shorter than `cdate`, like `d`, but I'd probably go with a more extreme model ([save those keystrokes](https://www.keysleft.com/)! 😅)

```shell
abbr c="cat \$(date +'%F').log"
c[ENTER] # expands to `cat $(date +'%F').log` and runs the command
```

**Reply by qadzek** [reply]
Thanks for your suggestions.

> Is there a reason you can't use the date command in the expansion itself?

I wasn't aware that this was possible. I tried, but it didn't work, possibly because I forgot to escape that `$` sign. That's why I ended up creating a clumsy Bash script :smiley:

I can't get `abbr cdate="\$(date +'%F')%ABBR_CURSOR%"` working. I'm not sure if I understand everything correctly, but I believe that I can't use the default `%` marker because it will conflict with the `%` in the `date` command. Changing `ABBR_LINE_CURSOR_MARKER` didn't seem to make a difference, modifying `ABBR_EXPANSION_CURSOR_MARKER` did. I opted for the `^` character. 

It looks like this works:

```zsh
# zshrc
ABBR_SET_EXPANSION_CURSOR=1
ABBR_EXPANSION_CURSOR_MARKER=^

# terminal
abbr -g cdate="\$(date +'%F')^"

**Reply by olets** [reply]
Nice!

> I'm not sure if I understand everything correctly, but I believe that I can't use the default % marker because it will conflict with the % in the date command

Exactly, without customizing the marker if you had

```
abbr -g cdate="\$(date +'%F')%"
```

then

```
cdate[SPACE]
```

would give you

```
$(date +'[CURSOR HERE]F')%
```

> Changing `ABBR_LINE_CURSOR_MARKER` didn't seem to make a difference, modifying `ABBR_EXPANSION_CURSOR_MARKER` did.

Hm `ABBR_EXPANSION_CURSOR_MARKER` defaults to `ABBR_LINE_CURSOR_MARKER` so setting just `ABBR_LINE_CURSOR_MARKER` _ought_ to set them both. But setting `ABBR_EXPANSION_CURSOR_MARKER` is clearer anyway 👍
